### PR TITLE
fix(Core/Movement): fix a small memory leak in RandomMovementGenerator

### PR DESCRIPTION
**Pull request by Grimdhex**
Thank to Ovahlord for the catch.

## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/20240

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ] Possibly don't fix the issue but the catch is correct
- [ ] 

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Grimdhex** [comment]
@PaoloBenigni could you test the PR ?

**Comment by PaoloBenigni** [comment]
On it

**Comment by walkline** [comment]
In general, refactoring it to use smart pointers is not a bad idea, but do you see a scenario where `_pathGenerator` could leak? I don't see one.

As for the crash, I don't think that check is necessary. The movement generator is bound to the creature's lifetime, so I would expect the pointer to be valid. Null cases should be handled at higher layers (if any).

As I mentioned in the Discord channel, the issue that @PaoloBenigni brought up is related to `_preComputedPaths`. It's a large cache, containing, I believe, 144 possible paths. Due to the randomness involved, filling in the cache takes a lot of time, and during this process, allocations are occurring.

**Comment by kissingers** [comment]
Comparing TC, found that there are a lot of differences,  the AZ does a lot of calculations on random moves, which may be the reason why the CPU load is higher.

https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp

**Comment by PaoloBenigni** [comment]
@walkline It seems like the calls are two types of object in memory, one related to line 72, and another in 137. The _precomputed is smaller, but there are many more hits than line 137.  Here are the relevant memory allocations. 


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Paolo/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Paolo/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


File | Line | Hits | Bytes | Bytes/hit
-- | -- | -- | -- | --
RandomMovementGenerator.h | 35 | 177 | 11328 | 64
RandomMovementGenerator.cpp | 72 | 254846 | 16310144 | 64
RandomMovementGenerator.cpp | 137 | 176 | 171776 | 976



</body>

</html>

**Comment by Grimdhex** [comment]
> In general, refactoring it to use smart pointers is not a bad idea, but do you see a scenario where `_pathGenerator` could leak? I don't see one.
> 
> 
> 
> As for the crash, I don't think that check is necessary. The movement generator is bound to the creature's lifetime, so I would expect the pointer to be valid. Null cases should be handled at higher layers (if any).
> 
> 
> 
> As I mentioned in the Discord channel, the issue that @PaoloBenigni brought up is related to `_preComputedPaths`. It's a large cache, containing, I believe, 144 possible paths. Due to the randomness involved, filling in the cache takes a lot of time, and during this process, allocations are occurring.

That why I note that will maybe don't fix the issue. But it's stay a safe case. We can't say that we expert with the management of our pointer (and the core code in general). Samething for the check. All of this cost nothing and avoid weird issues (or just mistake by our).

**Comment by PaoloBenigni** [comment]
Observations so far
image
Size of objects in memory is the same as before, It looks like there are fewer overall in memory. 
I poked around the world and didnt see any issues gameside.

**Comment by Grimdhex** [comment]
For _precuptedPath, honestly SetRandomPosition is so ugly that I'm not going to bother trying to figure out where it's stuck. I've imported TC's function, which is much cleaner and gets rid of a lot of variables. I've also made a few personal improvements to the rest of the class

I've tested 5min in games, and it doesn't seem to cause any crash.

**Comment by PaoloBenigni** [comment]
Ill pull and build now.

**Comment by walkline** [comment]
> Comparing TC, found that there are a lot of differences, the AZ does a lot of calculations on random moves, which may be the reason why the CPU load is higher.
> 
> https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp

> Comparing TC, found that there are a lot of differences, the AZ does a lot of calculations on random moves, which may be the reason why the CPU load is higher.
> 
> https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp

Your assumption regarding CPU consumption is wrong.

I compared speed of this TC port and the current implementation. Here is results:

| Name | median | average |
|---|---|---|
| This TC port | 58.292 µs |  70.297 µs |
| AC cache miss | 50.355 µs |  54.416 µs |
| AC cache hit | 15.125 µs |  16.182 µs |

The data is based on over 1000 calculations.

**Comment by PaoloBenigni** [comment]
I tested commit [37af537](https://github.com/azerothcore/azerothcore-wotlk/pull/20258/commits/37af537722e3018e6a5ba1147f84798fcc536dab) and found it to work. I teleported around and saw npcs moving around in what appeared a normal way.

**Comment by Grimdhex** [comment]
@walkine has absolutely right. TC doesn't seem to fire “less” calculation, it's just that they are attributed to other Core functions and not done directly in the method. So it's don't optimize the CPU calculation.

The only advantage with this PR is a more readable and understandable code and the removal of all those rather cumbersome variables that did seem to be the cause of this memory increase. But if someone feel to rewrite the class to optimize also the CPU usage, do it :)

**Comment by PaoloBenigni** [comment]
Tested latest. Built with VS. No observed issues.

**Comment by walkline** [comment]
> @walkine has absolutely right. TC doesn't seem to fire “less” calculation, it's just that they are attributed to other Core functions and not done directly in the method. So it's don't optimize the CPU calculation.
> 
> The only advantage with this PR is a more readable and understandable code and the removal of all those rather cumbersome variables that did seem to be the cause of this memory increase. But if someone feel to rewrite the class to optimize also the CPU usage, do it :)
> 
> Otherwise, ready to be review :)

I agree that the TC solution is easier to read and simple like bubble sort. The issue that @PaoloBenigni brought up doesn’t seem to be a problem, but rather intended behavior. It also appears that there’s a config option to disable the cache for the random movement generator if you’re limited on RAM.

Do we have any other issues with the current implementation? If not, then why consider changing the current efficient implementation to something less efficient, but easier to read (like bubble sort)?

**Comment by Grimdhex** [comment]
> > @walkine has absolutely right. TC doesn't seem to fire “less” calculation, it's just that they are attributed to other Core functions and not done directly in the method. So it's don't optimize the CPU calculation.
> 
> > 
> 
> > The only advantage with this PR is a more readable and understandable code and the removal of all those rather cumbersome variables that did seem to be the cause of this memory increase. But if someone feel to rewrite the class to optimize also the CPU usage, do it :)
> 
> > 
> 
> > Otherwise, ready to be review :)
> 
> 
> 
> I agree that the TC solution is easier to read and simple like bubble sort. The issue that @PaoloBenigni brought up doesn’t seem to be a problem, but rather intended behavior. It also appears that there’s a config option to disable the cache for the random movement generator if you’re limited on RAM.
> 
> 
> 
> Do we have any other issues with the current implementation? If not, then why consider changing the current efficient implementation to something less efficient, but easier to read (like bubble sort)?

@PaoloBenigni do you have test by disabling the option ?

**Comment by PaoloBenigni** [comment]
Ill test this now, but which part of the change is getting stuck with licensing issues?

**Comment by FrancescoBorzi** [comment]
@azerothcore/developers

File: src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp [hunk 1]
```diff
@@ -242,6 +241,8 @@ void RandomMovementGenerator<Creature>::DoInitialize(Creature* creature)
     if (!_wanderDistance)
         _wanderDistance = creature->GetWanderDistance();
 
+    _pathGenerator = nullptr;
```
Comment (Winfidonarleyan) [inline src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp:244]:
```suggestion
    _pathGenerator.reset();
```

File: src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp [hunk 2]
```diff
@@ -280,6 +281,7 @@ bool RandomMovementGenerator<Creature>::DoUpdate(Creature* creature, const uint3
     {
         _nextMoveTime.Reset(0);  // Expire the timer
         creature->StopMoving();
+        _pathGenerator = nullptr;
```
Comment (Winfidonarleyan) [inline src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp:284]:
```suggestion
        _pathGenerator.reset();
```

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by Ovahlord** [comment]
Don't use std::unique_ptr::reset(); like this... just go with = nullptr.

Also these reset calls are redundant and can be removed entirely. If you allocate a new object with std::make_unique, the previous object is automatically being destroyed and no longer needs manual management.

The pathGenerator class is doing nothing on its own so there is no need to destroy it unless you need a new one for a new path calculation. So tldr: kill the reset calls. Unnecessary memory meddling

**Pull request merged**